### DISCUSSION TITLE
How to achieve Message Priority using Quorum Queues?

### DISCUSSION THREAD

**Discussion by JacobPersi** [rabbitmq/rabbitmq-server#9516] [Questions]
My use case requires that RabbitMQ messages are processed (consumed) according to their priority. That is, high priority messages will be processed before medium priority, and medium before low.

According to the RabbitMQ documentation, this is possible with Quorum Queues using a queue for each priority.

Specifically, the RabbitMQ documentation states:

> Quorum queues support consumer priorities, but not message priorities. To prioritize messages with Quorum Queues, use multiple queues; one for each priority.

However, nothing more is said about how to implement this at the consumer level.

The only idea I have is to utilize the Rabbit "Polling" API to poll each queue in order, and only move on to a lower priority queue when no messages are returned; however, the documentation also says the Polling API is "highly inefficient and should be avoided in most cases".

Is there another approach to achieve message priorities using Quorum Queues?

**Comment by lukebakken** [comment]
* Consume from the highest priority queue. You will have to define a time interval that represents the longest time you'll wait for a message from this queue before moving to a lower priority one.
* If that time interval elapses, cancel your highest priority consumer, and start a consumer on the next lowest priority.
* Repeat the above until you have moved through all priorities, returning to the highest.

You have other options in this scenario, as well:
* Before moving to a lower priority queue, you could do a passive queue declaration on the higher priority queues just to check to see if they have messages. If so, make the highest priority with messages the queue from which to start the process.
* If you don't want to use a time interval to define "this priority is empty", you can use the passive queue declaration mechanism to check other queues to make decisions on where to consume from next. Perhaps you set your prefetch on a queue to 100 messages, and check other queues when each "batch" is done. It's up to you.

**Comment by michaelklishin** [comment]
Or use multiple consumers. Very often people reach for priority queues not because they absolutely must process message A before B but because they don't want a sequence of messages of type A to block processing of B for too long.

Multiple consumers on separate queues give you that without any polling or passive declaration dances.

You can also use an app-local priority queue where your consumers will store the deliveries, and process them concurrently however you like.

Arguably this would be much easier to reason about if instead of N queues you consider using N streams.
